amplify/backend/api/colonycdapp/schema.graphql

```diff
@@ -291,3 +310,27 @@ type ContractEvent @model {
   target: String! # who was on the receiving end of it
   encodedArguments: String # basically a JSON string
 }
+
+type ColonyExtension @model {
+  id: ID!
+  colonyId: ID!
+    @index(
+      name: "byColony"
+      sortKeyFields: ["hash"]
+      queryField: "getExtensionByColonyAndHash"
```

Comment:
What does the `ByColonyAndHash` mean here, as opposed to just `getExtensionByColony`?

Reply:
A single colony can have multiple extensions (or, more specifically, extension installations). This is also setting up an index on the extension hash for fast retrieval of a particular extension installation in a colony

Reply:
>I ask because at the moment we've got this distinction between extension and extensions

I believe this is a very much standard approach for any model - there are separate get all and get single queries:
<img width="164" alt="image" src="https://user-images.githubusercontent.com/112586815/209177556-f6eef2f7-86a4-48c7-b20f-86ca48ca8d14.png">

You could use one query and perform filtering on the client-side but I think the idea with the CDapp is to offload as much processing to the backend as possible. The benefits of having 2 separate queries are:
- Querying on the indexed fields (`byColonyAndHash`) is faster than filtering 
- When you're viewing an individual extension page, you can make a query to retrieve just the one extension instead of all of them
- Whenever dealing with `items`, Amplify will automatically introduce pagination and only return up to 100 items. Filtering happens after the pagination, so if the first item matching the filter is say 103rd in the database, it won't be returned by the query you proposed
While those might not be apparent at the start, I think they still outweigh the downsides.

Reply:
Do you think renaming `useExtensionData` to `useExtensionByIdData` would make the distinction clearer?

Reply:
@jakubcolony 

> I believe this is a very much standard approach for any model

Sorry, I was referring primarily to the hooks of similar name. 

Currently, I only see one main requirement:

1. get all installed extensions at a given colony id

From this, we can derive non-installed (i.e. available) extensions, and we can also derive whether a specific extension  is installed or not, thus my sense we only need one query. Maybe I could put my thoughts into code as it would be easier for you to review?

> Querying on the indexed fields (byColonyAndHash) is faster than filtering

When I run this query `getExtensionByColonyAndHash` in the graphql explorer, it gets all extensions installed at the given colony id, and then optionally if I provided an extension hash filters them by that hash. What is the difference between that and the query I wrote above? I understand the need to index the colonyId so we can search extensions at a given colony, but in both cases it's filtering that retrieves a specific extension at a specific colony? The website also alludes to `sortKeyField` as filtering: 

> [The example client query below allows you to query for "Customer" based on their name and filter based on phoneNumber:](https://docs.amplify.aws/cli/graphql/data-modeling/#configure-a-secondary-index)

My thinking is that if they are functionally equivaent, we can use the same query for both getting all extensions and getting a specific extension and simplfy/combine these two hooks.

Reply:
> Whenever dealing with items, Amplify will automatically introduce pagination and only return up to 100 items.

If I understand correctly per my comment above, we'd have the same problem with `getExtensionByColonyAndHash`?

Reply:
>Currently, I only see one main requirement:
>1. get all installed extensions at a given colony id

>From this, we can derive non-installed (i.e. available) extensions, and we can also derive whether a specific extension is installed or not, thus my sense we only need one query. Maybe I could put my thoughts into code as it would be easier for you to review?

I think it's fair to add a requirement of getting a single extension (with a specified hash) for a given colony. That way we can benefit of what I mentioned earlier:
>When you're viewing an individual extension page, you can make a query to retrieve just the one extension instead of all of them

I don't disagree this _could_ be done using a single query. Similarly to how we could call `listColonies` and filter them on the client for one with a given name instead of calling `getColonyByName`. But this is far from being optimal and it seems like all we're gaining is minimally simplified design.

From Amplify docs:
> Amazon DynamoDB is a key-value and document database that delivers single-digit millisecond performance at any scale but making it work for your access patterns requires a bit of forethought. DynamoDB query operations may use at most two attributes to efficiently query data. The first query argument passed to a query (the hash key) must use strict equality and the second attribute (the sort key) may use gt, ge, lt, le, eq, beginsWith, and between. DynamoDB can effectively implement a wide variety of access patterns that are powerful enough for the majority of applications.

Unfortunately they're quite useless but this suggests to me there is some underlying difference in how the data is stored and retrieved. 

This [SO thread](https://stackoverflow.com/questions/63226777/dynamodb-filter-and-sort-the-same-field) suggests to me the syntax of filtering and using sort key is the same (hence the understandable confusion), but the way filtering works is narrowing the entire partition of results whereas using sort key allows DynamoDB to select just what we're interested in. So it's not only faster, but we should not experience the same issue related to pagination if I understand this correctly.

Looks like there are a few ways to go about this (for example, we could use `getExtensionByColonyAndHash` without the `hash` parameter to retrieve all extensions in a given colony instead of getting them using Colony's `extensions` field) but I still believe there should be two queries - for all and a single extension, just like any other model and to ensure we're not querying more data than we need - it will at some point slow things down and cost more money.

Yes, we're probably talking distant future and/or small performance benefits but if the primary downside remains just slightly simpler design, I still think this is the right way to go

Reply:
Thanks for pointing me to the SO article, it in turn led me to the dynamo db docs which help explain [primary keys](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.CoreComponents.html#HowItWorks.CoreComponents.PrimaryKey), and I think you're right about preferring a sort key to a filter. 

> Looks like there are a few ways to go about this (for example, we could use getExtensionByColonyAndHash without  the hash parameter to retrieve all extensions in a given colony instead of getting them using Colony's extensions field)

I agree, but happy to leave as is for now since it's fairly minor.

Thanks for taking the time to have this discussion and help me understand dynamo db better. All good from me 👍